### ISSUE TITLE
批量预测问题

### ISSUE THREAD

**Issue by githubgtl** [zjunlp/KnowLM#97]
my code
import json
import os
import sys
from typing import List

import fire
import torch
import transformers
from datasets import load_dataset
from tqdm import tqdm

"""
Unused imports:
import torch.nn as nn
import bitsandbytes as bnb
"""

from peft import (
    LoraConfig,
    get_peft_model,
    get_peft_model_state_dict,
    prepare_model_for_int8_training,
    set_peft_model_state_dict,
)
from transformers import LlamaForCausalLM, LlamaTokenizer
from peft import PeftModel
from utils.prompter import Prompter

def train(
        # model/data params
        base_model: str = "/data/home/scv3183/knowlm",  # the only required argument
        data_path: str = "/data/home/scv3183/deepke/example/llm/finetune/lora/data/RE/valid.json",
        # training hyperparams
        batch_size: int = 128,
        micro_batch_size: int = 4,
        num_epochs: int = 3,
        learning_rate: float = 3e-4,
        cutoff_len: int = 512,
        val_set_size: int = 2000,
        train_on_inputs: bool = False,  # if False, masks out inputs in loss
        group_by_length: bool = False,  # faster, but produces an odd training loss curve
        # wandb params
        wandb_project: str = "",
        wandb_run_name: str = "",
        wandb_watch: str = "",  # options: false | gradients | all
        wandb_log_model: str = "",  # options: false | true
        # either training checkpoint or final adapter
        prompt_template_name: str = "alpaca",  # The prompt template to use, will default to alpaca.
        # train hyperparams
        warmup_steps=10,
        logging_steps=1,
        save_steps=10,
        save_total_limit=5,
        eval_steps=10,
):
    base_model = "/data/home/scv3183/knowlm"
    lora_weights = "../../finetune/lora/checkpoint/checkpoint-10"
    tokenizer = LlamaTokenizer.from_pretrained(base_model)
    output_dir = "../results/knowlm-valid-ori-1.json"
    from tqdm import tqdm
    prompter = Prompter(prompt_template_name)
    use_wandb = len(wandb_project) > 0 or (
            "WANDB_PROJECT" in os.environ and len(os.environ["WANDB_PROJECT"]) > 0
    )
    # Only overwrite environ if wandb param passed
    if len(wandb_project) > 0:
        os.environ["WANDB_PROJECT"] = wandb_project
    if len(wandb_watch) > 0:
        os.environ["WANDB_WATCH"] = wandb_watch
    if len(wandb_log_model) > 0:
        os.environ["WANDB_LOG_MODEL"] = wandb_log_model
    world_size = int(os.environ.get("WORLD_SIZE", 1))
    ddp = world_size != 1

    if torch.cuda.is_available():
        print("device is cuda")
        device = "cuda"
    else:
        device = "cpu"

    try:
        if torch.backends.mps.is_available():
            device = "mps"
    except:
        pass
    if device == "cuda":
        model = LlamaForCausalLM.from_pretrained(
            base_model,
            load_in_8bit=False,
            torch_dtype=torch.float16,
            device_map="auto",
        )
        model = PeftModel.from_pretrained(
            model,
            lora_weights,
            torch_dtype=torch.float16,
        )
        model.config.pad_token_id = tokenizer.pad_token_id = 0  # same as unk token id
        model.config.bos_token_id = tokenizer.bos_token_id = 1
        model.config.eos_token_id = tokenizer.eos_token_id = 2
        print("loaded model")
    model.eval()
    if torch.__version__ >= "2" and sys.platform != "win32":
        model = torch.compile(model)

    def tokenize(prompt, add_eos_token=True):
        # there's probably a way to do this with the tokenizer settings
        # but again, gotta move fast
        result = tokenizer(
            prompt,
            truncation=True,
            max_length=cutoff_len,
            padding=False,
            return_tensors=None,
        )
        if (
                result["input_ids"][-1] != tokenizer.eos_token_id
                and len(result["input_ids"]) < cutoff_len
                and add_eos_token
        ):
            result["input_ids"].append(tokenizer.eos_token_id)
            result["attention_mask"].append(1)

        result["labels"] = result["input_ids"].copy()

        return result

    def generate_and_tokenize_prompt(data_point):
        full_prompt = prompter.generate_prompt(
            data_point["instruction"],
            data_point["input"],
            data_point["output"],
        )
        tokenized_full_prompt = tokenize(full_prompt)
        if not train_on_inputs:
            user_prompt = prompter.generate_prompt(
                data_point["instruction"], data_point["input"]
            )
            tokenized_user_prompt = tokenize(user_prompt, add_eos_token=False)
            user_prompt_len = len(tokenized_user_prompt["input_ids"])

            tokenized_full_prompt["labels"] = [
                                                  -100
                                              ] * user_prompt_len + tokenized_full_prompt["labels"][
                                                                    user_prompt_len:
                                                                    ]  # could be sped up, probably
        return tokenized_full_prompt

    if data_path.endswith(".json") or data_path.endswith(".jsonl"):
        data = load_dataset("json", data_files=data_path)
        print(f"data includes: {data_path}")
    else:
        """is folder"""
        data_paths = []
        data_path = data_path if data_path[-1] == "/" else data_path + "/"
        for i in os.listdir(data_path):
            data_paths.append(os.path.join(data_path, i))
        print(f"data includes: {data_paths}")
        data = load_dataset(data_paths)

    if val_set_size > 0:
        test_data = (
            data["train"].shuffle().map(generate_and_tokenize_prompt)
        )

    gradient_accumulation_steps = batch_size // micro_batch_size
    if ddp:
        gradient_accumulation_steps //= world_size
    trainer = transformers.Trainer(
        model=model,
        args=transformers.TrainingArguments(
            per_device_train_batch_size=micro_batch_size,
            gradient_accumulation_steps=gradient_accumulation_steps,
            warmup_steps=warmup_steps,
            num_train_epochs=num_epochs,
            learning_rate=learning_rate,
            fp16=True,
            logging_steps=logging_steps,
            optim="adamw_torch",
            evaluation_strategy="steps" if val_set_size > 0 else "no",
            save_strategy="steps",
            eval_steps=eval_steps if val_set_size > 0 else None,
            save_steps=save_steps,
            output_dir="output_dir",
            save_total_limit=save_total_limit,
            load_best_model_at_end=True if val_set_size > 0 else False,
            ddp_find_unused_parameters=False if ddp else None,
            group_by_length=group_by_length,
            report_to="wandb" if use_wandb else None,
            run_name=wandb_run_name if use_wandb else None,
        ),
        data_collator=transformers.DataCollatorForSeq2Seq(
            tokenizer, pad_to_multiple_of=8, return_tensors="pt", padding=True
        ),
    )
    predictions = trainer.predict(test_data)
    records = []
    with open(data_path, "r") as reader:
        for line in reader:
            data = json.loads(line)
            records.append(data)

    with open(output_dir, "w") as writer:
        for i, (record, result) in tqdm(enumerate(zip(records, predictions))):
            print(i, result)
            record['output'] = result
            writer.write(json.dumps(record, ensure_ascii=False) + '\n')
    # model.save_pretrained(output_dir)

if __name__ == "__main__":
    fire.Fire(train)
得到的信息不是预测信息，我有单个预测的代码，但那太慢了

**Comment by MikeDean2367** [comment]
您好，我没理解您的意思，从您提供的代码来看是训练代码，但是从您的描述来看是推理。

**Comment by MikeDean2367** [comment]
关于批量预测，`transformers`包中的`generate()`方法本身支持批量预测。

首先需要加载模型和分词器得到`model`和`tokenizer`变量，然后将您的批量输入送入到分词器进行分词，这样您就得到了`batch_size*length`形状的向量，最后将其传入到`model.generate()`即可。

**Comment by githubgtl** [comment]
不是训练的代码，我希望使用trainer.predict函数进行批量预测，但是发现得到的不是我想要的预测值



---原始邮件---
发件人: "Mike ***@***.***&gt;
发送时间: 2023年11月4日(周六) 下午2:17
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [zjunlp/KnowLM] 批量预测问题 (Issue #97)




 
您好，我没理解您的意思，从您提供的代码来看是训练代码，但是从您的描述来看是推理。

**Comment by githubgtl** [comment]
谢谢您的分享，祝您心想事成，万事如意



---原始邮件---
发件人: "Mike ***@***.***&gt;
发送时间: 2023年11月4日(周六) 下午4:35
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [zjunlp/KnowLM] 批量预测问题 (Issue #97)




 
关于批量预测，transformers包中的generate()方法本身支持批量预测。
 
首先需要加载模型和分词器得到model和tokenizer变量，然后将您的批量输入送入到分词器进行分词，这样您就得到了batch_size*length形状的向量，最后将其传入到model.generate()即可。

**Comment by zxlzr** [comment]
请问您的问题是否已解决？

**Comment by githubgtl** [comment]
> 请问您的问题是否已解决？
我可以通过data.map()得到list【input_ids】，但是直接输入到model。generate（会有维度错误
）

**Comment by MikeDean2367** [comment]
您好，您没有提供代码，下面是我提供的伪代码：
```python
model = LlamaForCausalLM.from_pretrained("model_name")
tokenizer = LlamaTokenizer.from_pretrained("model_name")
# 设置分词器的padding token
inputs = [
     "input1",
     "input3"
]
input_ids = tokenizer(inputs, return_tensors="pt",  padding=True)['input_ids']
output = model.generate(input_ids=input_ids, max_new_tokens=10)
ans = [tokenizer.decode(x) for x in outputs.detach().cpu().numpy().tolist()]
print(ans)
```
此时模型将并行处理2条输入（请确保有足够的显存，更多的解码参数可以在`model.generate()`中传入）。如有问题请告知我 :)

**Comment by githubgtl** [comment]
import json
import os
import sys
from typing import List

import fire
import torch
import transformers
from datasets import load_dataset
from tqdm import tqdm

"""
Unused imports:
import torch.nn as nn
import bitsandbytes as bnb
"""

from peft import (
    LoraConfig,
    get_peft_model,
    get_peft_model_state_dict,
    prepare_model_for_int8_training,
    set_peft_model_state_dict,
)
from transformers import LlamaForCausalLM, LlamaTokenizer, GenerationConfig
from peft import PeftModel
from utils.prompter import Prompter


def train(
        base_model: str = "/data/home/scv3183/knowlm",  # the only required argument
        data_path: str = "/data/home/scv3183/deepke/example/llm/finetune/lora/data/RE/valid.json",
        cutoff_len: int = 512,
        val_set_size: int = 2000,
        train_on_inputs: bool = False,  # if False, masks out inputs in loss
        prompt_template_name: str = "alpaca",  # The prompt template to use, will default to alpaca.
):
    lora_weights = "../../finetune/lora/checkpoint/checkpoint-10"
    tokenizer = LlamaTokenizer.from_pretrained(base_model)
    output_dir = "../results/knowlm-valid-ori-1.json"
    from tqdm import tqdm
    prompter = Prompter(prompt_template_name)
    if torch.cuda.is_available():
        print("device is cuda")
        device = "cuda"
    else:
        device = "cpu"

    try:
        if torch.backends.mps.is_available():
            device = "mps"
    except:
        pass
    def tokenize(prompt, add_eos_token=True):
        """

        :param prompt:
        :param add_eos_token:
        :return:
        """
        return tokenizer(
            prompt,
            return_tensors="pt",
        )

    def generate_and_tokenize_prompt(data_point):
        full_prompt = prompter.generate_prompt(
            data_point["instruction"],
            data_point["input"],
        )
        tokenized_full_prompt = tokenize(full_prompt)
        return tokenized_full_prompt

    if data_path.endswith(".json") or data_path.endswith(".jsonl"):
        data = load_dataset("json", data_files=data_path)
        print(f"data includes: {data_path}")
    else:
        """is folder"""
        data_paths = []
        data_path = data_path if data_path[-1] == "/" else data_path + "/"
        for i in os.listdir(data_path):
            data_paths.append(os.path.join(data_path, i))
        print(f"data includes: {data_paths}")
        data = load_dataset(data_paths)

    if val_set_size > 0:
        test_data = data.map(generate_and_tokenize_prompt)
    input_ids = test_data.data['train']['input_ids']

    if device == "cuda":
        model = LlamaForCausalLM.from_pretrained(
            base_model,
            load_in_8bit=False,
            torch_dtype=torch.float16,
            device_map="auto",
        )
        model = PeftModel.from_pretrained(
            model,
            lora_weights,
            torch_dtype=torch.float16,
        )
        model.config.pad_token_id = tokenizer.pad_token_id = 0  # same as unk token id
        model.config.bos_token_id = tokenizer.bos_token_id = 1
        model.config.eos_token_id = tokenizer.eos_token_id = 2
        print("loaded model")
    model.eval()
    if torch.__version__ >= "2" and sys.platform != "win32":
        model = torch.compile(model)


    generation_config = GenerationConfig(
        temperature=1.0,
        top_p=1.0,
        top_k=50,
        num_beams=1,
        max_new_tokens=256,
        pad_token_id=tokenizer.pad_token_id,
        return_dict_in_generate=True,
        output_scores=True
    )
    with torch.no_grad():
        outputs = model.generate(input_ids=input_ids, generation_config=generation_config)

    # 解码生成的文本
    decoded_texts = [tokenizer.decode(output.sequences[0], skip_special_tokens=True) for output in outputs]
    with open(output_dir, "w") as writer:
        for i, result in tqdm(enumerate(decoded_texts)):
            print(i, result)
            writer.write(result + '\n')


if __name__ == "__main__":
    fire.Fire(train)感觉是一样的逻辑

**Comment by MikeDean2367** [comment]
您好，请按照我之前提供的代码进行修改。

**Comment by githubgtl** [comment]
> ```python
> input_ids = tokenizer(inputs, return_tensors="pt",  padding=True)['input_ids']
> ```

![image](https://github.com/zjunlp/KnowLM/assets/88779693/49419f6d-926e-48d6-bb3b-c9a6482edee6)

**Comment by MikeDean2367** [comment]
您好，请使用如下代码（经过测试可用）：
```python
from transformers import LlamaForCausalLM, LlamaTokenizer, GenerationConfig
import torch

device = "cuda"
repo = "llama2-7b-chat"
model = LlamaForCausalLM.from_pretrained(repo, device_map="auto", torch_dtype=torch.bfloat16)
tokenizer = LlamaTokenizer.from_pretrained(repo)
tokenizer.pad_token_id = tokenizer.unk_token_id
tokenizer.padding_side='left'

inputs = [
     "who are you?",
     "do you know zhejiang university?"
]
input_ids = tokenizer(inputs, return_tensors="pt",  padding=True)['input_ids'].to(device)
outputs = model.generate(input_ids=input_ids, max_new_tokens=100)
ans = [tokenizer.decode(x) for x in outputs.detach().cpu().numpy().tolist()]
print(ans)
```
请修改里面的`repo`变量，此外如果您的显卡支持`bf16`，则保持`torch_dtype=torch.bfloat16`，否则修改成`torch.float16`或`torch.float32`。

如有其他问题，请告知我 :)

**Comment by zxlzr** [comment]
请问您的问题是否已解决

**Comment by githubgtl** [comment]
> 请问您的问题是否已解决

已解决
